scalyr_agent/util.py

```diff
@@ -55,47 +55,82 @@
 except ImportError:
     uuid = None
 
-def _fallback_json_encode( obj ):
+
+def _fallback_json_encode( obj, sort_keys=False ):
+    # json_lib.serialize() always ignores sort_keys param (always sorts regardless)
     return json_lib.serialize( obj )
 
+
 def _fallback_json_decode( text ):
     return json_lib.parse( text )
 
-_json_lib = 'json_lib'
 
-_json_encode = _fallback_json_encode
-_json_decode = _fallback_json_decode
+def get_json_implementation(lib_name):
 
-try:
-    import ujson
-    _json_lib = 'ujson'
-    _json_encode = ujson.dumps
-    _json_decode = ujson.loads
-except ImportError:
-    try:
+    if lib_name == 'json_lib':
+        return lib_name, _fallback_json_encode, _fallback_json_decode
+
+    elif lib_name == 'ujson':
+        import ujson
+
+        def ujson_dumps_custom(*args, **kwargs):
+            # ujson does not raise exception if you pass it a JsonArray/JsonObject while producing wrong encoding.
+            # Detect and complain loudly.
+            if isinstance(args[0], (json_lib.JsonObject, json_lib.JsonArray)):
+                raise TypeError('ujson does not correctly encode objects of type: %s' % type(args[0]))
+            return ujson.dumps(*args, **kwargs)
+
+        return lib_name, ujson_dumps_custom, ujson.loads
+
+    if lib_name == 'json':
         import json
-        _json_lib = 'json'
 
-        def dumps_no_space(*args, **kwargs):
+        def json_dumps_custom(*args, **kwargs):
             """Eliminate spaces by default. Python 2.4 does not support partials."""
             if 'separators' not in kwargs:
                 kwargs['separators'] = (',', ':')
             return json.dumps(*args, **kwargs)
 
-        _json_encode = dumps_no_space
-        _json_decode = json.loads
+        return lib_name, json_dumps_custom, json.loads
+
```

Comment:
If someone passes in a `lib_name` value that is not `json` `ujson` or `json_lib` (for whatever reason, maybe a spelling mistake, or maybe a new json library is added), then the function will reach the end and return `None`, instead of an expected 3-value tuple.

It's not a big issue, because it'll probably never happen and if it does we'll get a compile error, but maybe if we get an unrecognized name then we raise an exception with a message that the json `lib_name` is unsupported.

Reply:
Thanks, good suggestion.  Change has been made & unit test installed to capture.